infrastructure/src/main/java/org/corfudb/infrastructure/datastore/DataStore.java

```diff
@@ -129,12 +133,19 @@ public synchronized void write(@Nonnull String key, @Nonnull Object value) {
                             String jsonPayload = JsonUtils.parser.toJson(value, value.getClass());
                             byte[] bytes = jsonPayload.getBytes();
 
-                            ByteBuffer buffer = ByteBuffer.allocate(bytes.length
-                                    + Integer.BYTES);
-                            buffer.putInt(getChecksum(bytes));
-                            buffer.put(bytes);
-                            Files.write(tmpPath, buffer.array(), StandardOpenOption.CREATE,
+                            buffer = Unpooled.directBuffer(bytes.length + Integer.BYTES);
+                            buffer.writeInt(getChecksum(bytes));
+                            buffer.writeBytes(bytes);
```

Comment:
To eliminate the other buffer copy you can just use `buffer.writeCharSequence(jsonPayload, Charset.defaultCharset());` , or maybe explicitly set it to UTF-8

Reply:
There are some implementation problems if we do buffer.writeCharSequence. 1) depending on the charset, the final capacity of byte buffer is usually not equal to string.length(), so there might be overheads to scale the the initial capacity. 2) the  checksum needs to be serialized before the payload, so we need to rewind the index and rewrite checksum, and the hasher.putBytes(ByteBuffer bytes) takes a nio buffer instead of a netty bytebuf, which requires more index manipulation, seems a bit unclean and error prone...

Reply:
Its actually not variable for `Layout.java`, it can be fixed to UTF-8. You're already using a direct buffer.